### ISSUE TITLE
Embind: instantiate multiple times

### ISSUE THREAD

**Issue by ivan-aksamentov** [emscripten-core/emscripten#14978]
I want Embind to allow instantiating the same module multiple times., i.e.:
 - download once
 - compile once
 - instantiate multiple independent modules

This would be something along these lines with vanilla `WebAssembly.instantiateStreaming()` + `WebAssembly.instantiate()`:

```js
const fetchPromise = fetch('fibonacci.wasm');
const { module, instance } = await WebAssembly.instantiateStreaming(fetchPromise);
const otherInstance = await WebAssembly.instantiate(module);
```
(taken from [developers.google.com: Loading WebAssembly modules efficiently](https://developers.google.com/web/updates/2018/04/loading-wasm))

Note the `instance` and `otherInstance` are the 2 instances (I need even more) of the same `module`.

What I do right now is horribly inefficient and users have to wait a few seconds for the entire thing to complete. Right now I am forced (I think) to download, compile and instantiate in one go:

```js
  return new Promise((resolve) => {
    const module = emscriptenJs({
      onRuntimeInitialized: () => {
        resolve(module)
      },
    })
  })
```

which means doing a lot of useless duplicated work multiple times.

(source: (1) [the loop calling init function on WebWorker](https://github.com/nextstrain/nextclade/blob/34d262d3327267833e55629d6ee0df0061301088/packages/web/src/workers/run.ts#L39-L44), (2) [the init function calling loading code](https://github.com/nextstrain/nextclade/blob/34d262d3327267833e55629d6ee0df0061301088/packages/web/src/workers/worker.analyze.ts#L73-L88), (3) [the loading code](https://github.com/nextstrain/nextclade/blob/34d262d3327267833e55629d6ee0df0061301088/packages/web/src/workers/wasmModule.ts#L47-L59))

Context: I am using a WebWorker thread pool, where each WebWorker does the same thing, and the work is implemented in C++/WebAssembly. My idea is to compile once, pass the compiled blob to WebWorkers, instantiate there.

Is it currently possible to instantiate Embind-based modules multiple times? So far I tried to instantiate them with the vanilla code above, but they don't have imports section and even if they did, will probably require the JS glue code.

If it's not yet possible, I am ready to try to implement a solution, either in userland or in Emscripten, so I appreciate any hints.

**Comment by sbc100** [comment]
We have a hook in emscripten for supplying your own pre-compiled module: https://emscripten.org/docs/api_reference/module.html#Module.instantiateWasm.

I think this may provide what you need?

**Comment by ivan-aksamentov** [comment]
@sbc100 Thanks Sam, 

This callback looks like a place where I can put the vanilla `WebAssembly` `instantiateStreaming()` and `instantiate()` calls. I will give it a try.

I guess my confusion is based on the fact that the Embind module interface is a bit backwards. It tries to do some hidden magic and exposes callbacks to customize some of it, instead of simply allowing calling functions the way I want, like in vanilla `WebAssembly`, perhaps being a wrapper around them. But I don't actually know why that particular design was chosen, maybe it makes sense internally.

Thanks again for the suggestion!

**Comment by kripken** [comment]
Would the MODULARIZE option help here? See e.g. https://emscripten.org/docs/porting/connecting_cpp_and_javascript/Interacting-with-code.html?highlight=modularize#interacting-with-an-api-written-in-c-c-from-nodejs

**Comment by renspoesse** [comment]
@ivan-aksamentov did you find any solution to this? I'm trying to do the same thing.

**Comment by ivan-aksamentov** [comment]
@renspoesse No. But the `Module.instantiateWasm()` Sam mentioned above seems promising. I just never had enough time to dig into it. Let me know if you find out something, as I am still interested.
